### PR TITLE
Integration  check script fix ups

### PR DESCRIPTION
**Pull request by maxknv**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix integration check python script to use gh api url
- Add Readme for CI tests

### Documentation entry for user-facing changes

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 0b085f459c4df85acf2c018e86e027a985c9775b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55476/0b085f459c4df85acf2c018e86e027a985c9775b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6471966469">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55476/0b085f459c4df85acf2c018e86e027a985c9775b/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: tests/ci/pr_info.py [hunk 1]
```diff
@@ -264,16 +264,11 @@ def fetch_changed_files(self):
...
             response.raise_for_status()
-            if "commits" in self.event and self.number == 0:
-                diff = response.json()
-
-                if "files" in diff:
-                    self.changed_files = {f["filename"] for f in diff["files"]}
-            else:
-                diff_object = PatchSet(response.text)
-                self.changed_files.update({f.path for f in diff_object})
+            diff_object = PatchSet(response.text)
```
Comment (Felixoid) [inline tests/ci/pr_info.py:270]: Let's remove patchset and use only `compare/ref1...ref2` json. And remove unidiff from https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/worker/prepare-ci-ami.sh#L93C45-L93C52

File: tests/ci/build_download_helper.py [hunk 2]
```diff
@@ -89,6 +89,7 @@ def set_auth_header():
                 set_auth_header()
                 need_retry = True
                 break
+            raise exc
```
Comment (Felixoid) [inline tests/ci/build_download_helper.py:92]: Accidentally added in https://github.com/ClickHouse/ClickHouse/commit/1fbfcbc3d3f9d7bd047d8ae2da8b658c28249e15#r129706939

It neglects `for _ in range(retries):` from line 75, so it shouldn't be here

Comment (Felixoid) [reply tests/ci/build_download_helper.py:92]:
```
In [3]: for i in range(5):
   ...:     print(i)
   ...: else:
   ...:     print('finished')
   ...:
0
1
2
3
4
finished

In [4]: for i in range(5):
   ...:     print(i)
   ...:     if i > 3:
   ...:         break
   ...: else:
   ...:     print('finished')
   ...:
0
1
2
3
4
```

It looks like we can use `else` here

File: tests/ci/worker/prepare-ci-ami.sh [hunk 3]
```diff
@@ -90,7 +90,7 @@ systemctl restart docker
 sudo -u ubuntu docker buildx version
 sudo -u ubuntu docker buildx create --use --name default-builder
 
-pip install boto3 pygithub requests urllib3 unidiff dohq-artifactory
+pip install boto3 pygithub requests urllib3 dohq-artifactory
```
Comment (Felixoid) [inline tests/ci/worker/prepare-ci-ami.sh:93]: It's in the Dockerfile too

```
git grep -n unidiff
docker/test/style/Dockerfile:22:    && pip3 install black==23.1.0 boto3 codespell==2.2.1 mypy==1.3.0 PyGithub unidiff pylint==2.6.2 \
```

File: tests/README.md
Comment (tavplubix) [inline tests/README.md:1]: Consider adding all this stuff to https://clickhouse.com/docs/en/development/continuous-integration instead and keeping only a link to the documentation in this README

File: docs/en/development/continuous-integration.md [hunk 4]
```diff
@@ -67,6 +67,51 @@ This check means that the CI system started to process the pull request. When it
 Performs some simple regex-based checks of code style, using the [`utils/check-style/check-style`](https://github.com/ClickHouse/ClickHouse/blob/master/utils/check-style/check-style) binary (note that it can be run locally).
 If it fails, fix the style errors following the [code style guide](style.md).
 
+#### Running localy in docker:
+```sh
+mkdir /tmp/test_output
```
Comment (Felixoid) [inline docs/en/development/continuous-integration.md:72]:
```suggestion
mkdir -p /tmp/test_output
```

to suppress `already exists`

File: docs/en/development/continuous-integration.md [hunk 5]
```diff
@@ -75,6 +120,77 @@ some. If it fails, further checks are not started until it is fixed. Look at
 the report to see which tests fail, then reproduce the failure locally as
 described [here](tests.md#functional-test-locally).
 
+Fast test job builds minimal CH, if you for watever reason need the same ClickHouse build, follow the instructions:
+```sh
```
Comment (Felixoid) [inline docs/en/development/continuous-integration.md:124]: I feel it's better to use fast-tests image

File: docs/en/development/continuous-integration.md [hunk 6]
```diff
@@ -122,6 +238,27 @@ Builds ClickHouse in various configurations for use in further steps. You have t
 ## Special Build Check
 Performs static analysis and code style checks using `clang-tidy`. The report is similar to the [build check](#build-check). Fix the errors found in the build log.
 
+#### Running localy in docker:
```
Comment (Felixoid) [inline docs/en/development/continuous-integration.md:241]: https://clickhouse.com/docs/en/development/build#building-in-docker

better to give this link here

File: tests/ci/build_download_helper.py [hunk 7]
```diff
@@ -71,27 +71,38 @@ def set_auth_header():
...
+            exc = e
+            ratelimit_exceeded = (
+                e.response.status_code == 403
                 and b"rate limit exceeded"
-                in exc.response._content  # pylint:disable=protected-access
-            ):
+                in e.response._content  # pylint:disable=protected-access
+            )
+            try_auth = e.response.status_code == 404
+            if (ratelimit_exceeded or try_auth) and not token_is_set:
```
Comment (Felixoid) [inline tests/ci/build_download_helper.py:91]: we need to update `token_is_set` in the loop

File: docker/test/style/Dockerfile [hunk 8]
```diff
@@ -19,7 +19,7 @@ RUN apt-get update && env DEBIAN_FRONTEND=noninteractive apt-get install --yes \
     shellcheck \
     yamllint \
     locales \
-    && pip3 install black==23.1.0 boto3 codespell==2.2.1 mypy==1.3.0 PyGithub unidiff pylint==2.6.2 \
```
Comment (Felixoid) [inline docker/test/style/Dockerfile:22]: You won't believe.. but we need bring it back, and leave "FIXME" comment here to get off rid this package in a couple of months. As well as in Ami script.

Old PRs will break horribly.

Besides that, we need to backport the changes everywhere.

**Review by alesapin — APPROVED** [review]

**Review by Felixoid — APPROVED** [review]

File: tests/ci/pr_info.py [hunk 9]
```diff
@@ -171,7 +169,11 @@ def __init__(
                     response_json = user_orgs_response.json()
                     self.user_orgs = set(org["id"] for org in response_json)
 
-            self.diff_urls.append(github_event["pull_request"]["diff_url"])
+            self.diff_urls.append(
+                f"https://api.github.com/repos/{GITHUB_REPOSITORY}/"
+                f"compare/master...{self.head_ref}"
```
Comment (Felixoid) [inline tests/ci/pr_info.py:174]: I missed it at the review. We need to use `github_event["pull_request"]["head"]["label"]` here

File: tests/ci/pr_info.py [hunk 10]
```diff
@@ -215,12 +217,12 @@ def __init__(
...
-                        f"https://github.com/{GITHUB_REPOSITORY}/"
-                        f"compare/master...{self.head_ref}.diff"
+                        f"https://api.github.com/repos/{GITHUB_REPOSITORY}/"
+                        f"compare/master...{self.head_ref}"
                     )
                     self.diff_urls.append(
-                        f"https://github.com/{GITHUB_REPOSITORY}/"
-                        f"compare/{self.head_ref}...master.diff"
+                        f"https://api.github.com/repos/{GITHUB_REPOSITORY}/"
+                        f"compare/{self.head_ref}...master"
```
Comment (Felixoid) [inline tests/ci/pr_info.py:225]: And here and below we need to use `pull_request["head"]["label"]`

Comment (mkaynov) [reply tests/ci/pr_info.py:225]: Yes. I wonder why self.head_ref worked before here.

**Pull request merged**